system/src/system_cloud_internal.cpp

```diff
@@ -1122,6 +1124,42 @@ inline uint8_t spark_cloud_socket_closed()
     return closed;
 }
 
+int formatOtaUpdateStatusEventData(uint32_t flags, int result, hal_module_t* module, uint8_t *buf, size_t size)
+{
+    int res = 1;
+    memset(buf, 0, size);
+
+    BufferAppender appender(buf, size);
+    appender.append("{");
+    appender.append("\"r\":");
+    appender.append(result ? "\"error\"" : "\"ok\"");
+
+    if (flags & 1) {
+        appender.append(",");
+        res = ota_update_info(append_instance, &appender, module, false, NULL);
+    }
+
+    appender.append("}");
+
+    return res;
+}
+
+int finish_ota_firmware_update(FileTransfer::Descriptor& file, uint32_t flags, void*)
+{
+    hal_module_t module;
+    uint8_t buf[512];
+
+    int result = Spark_Finish_Firmware_Update(file, flags, &module);
+
+    formatOtaUpdateStatusEventData(flags, result, &module, buf, sizeof(buf));
+
+    Particle.publish("spark/device/ota_result", (const char*)buf, 60, PRIVATE);
+    HAL_Delay_Milliseconds(1000);
```

Comment:
I'm assuming we need this to ensure the publish is sent before we soft reset to apply the OTA update.  Can we make this more synchronous, as implemented in PR #1144 ?  We can do this as a separate PR after these get merged.  I'll take notes on all of these review comments to make sure we don't forget :)